python/dgl/data/csv_dataset.py

```diff
@@ -0,0 +1,420 @@
+                    data[key] = F.tensor(value[idx])
+                edata['data'] = data
+                if graph_id not in edge_dict:
+                    edge_dict[graph_id] = {}
+                edge_dict[graph_id][e_data.type] = edata
+        return edge_dict
+
+    @staticmethod
+    def _parse_graph_data(graph_data, graphs_dict):
+        if len(graphs_dict) > len(graph_data.graph_id):
+            raise DGLError(
+                "More graph ids are found in node/edge data than graph data. Please specify all graph ids in graph data CSV.")
+        graph_ids = graph_data.graph_id
+        graphs = []
+        for graph_id in graph_ids:
+            if graph_id not in graphs_dict:
+                graphs_dict[graph_id] = dgl_heterograph(
+                    {('_V', '_E', '_V'): ([], [])})
+        for graph_id in graph_ids:
+            graphs.append(graphs_dict[graph_id])
+        data = {}
+        for k, v in graph_data.data.items():
+            data[k] = F.tensor(v)
+        return graphs, data
+
+    @staticmethod
+    def _construct_graphs(node_dict, edge_dict):
+        graph_dict = {}
+        for graph_id in node_dict:
+            if graph_id not in edge_dict:
+                edata = {}
+                edata['edges'] = ([], [])
+                edge_dict[graph_id][('_V', '_E', '_V')] = edata
+
+            edges = {}
+            for etype, edata in edge_dict[graph_id].items():
+                edges[etype] = edata['edges']
+            nodes = {}
+            for ntype, ndata in node_dict[graph_id].items():
+                nodes[ntype] = len(ndata['mapping'])
+            graph = dgl_heterograph(
+                edges, num_nodes_dict=nodes)
+
+            def assign_data(type, src_data, dst_data):
+                for key, value in src_data.items():
+                    dst_data[type].data[key] = value
+            for type, data in node_dict[graph_id].items():
+                assign_data(type, data['data'], graph.nodes)
+            for (type), data in edge_dict[graph_id].items():
+                assign_data(type, data['data'], graph.edges)
+            graph_dict[graph_id] = graph
+        return graph_dict
+
+
+class DefaultDataParser:
+    """ Default data parser for DGLCSVDataset. It
+        1. ignores any columns which does not have a header.
+        2. tries to convert to list of numeric values if cell data is a str.
+        3. read data and infer data type directly, otherwise.
+    """
+
+    def __init__(self, separator='|'):
+        self.separator = separator
+
+    def __call__(self, df):
+        data = {}
+        for header in df:
+            if 'Unnamed' in header:
+                dgl_warning("Unamed column is found. Ignored...")
+                continue
+            dt = df[header].to_numpy().squeeze()
+            if isinstance(dt[0], str):
+                #probably consists of list of numeric values
+                dt = np.array([ast.literal_eval(row) for row in dt])
```

Comment:
`self.separator` is not used?

Reply:
I removed such separator from YAML as I find it's not intuitive and simple to parse such data into array. Though we could split it manually, this mainly depends on how user generate such list data. For simplicity, we could ask user use `np.array().tolist()` and parse with `ast.literal_eval`. Or the list data could be "[1,2,3]", 1|2|3 or any other forms

Reply:
In this case, we probably should just say we only support "1,2,3" at the moment.

Reply:
you mean: we only support "1,2,3", and split on our own instead of `ast.literal_eval`?

Reply:
You could still use literal_eval for implementation but we won't allow users to specify a different separator.

Reply:
ok.